compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -2240,6 +2240,174 @@ OMR::Power::TreeEvaluator::lbitpermuteEvaluator(TR::Node *node, TR::CodeGenerato
+         {
+         int64_t src = srcNode->getUnsignedLongInt() & zeroExtendMask;
+         srcReg = cg->allocateRegister();
+         loadConstant(cg, srcNode, src, srcReg);
+         resReg = srcReg;
+         }
+      // Otherwise check if either is in memory and can therefore be loaded with a zero-extending opcode
+      else if (maskNode->getReferenceCount() == 1 && maskNode->getOpCode().isMemoryReference() && maskNode->getRegister() == NULL)
+         {
+         maskReg = cg->allocateRegister();
+         TR::LoadStoreHandler::generateLoadNodeSequence(cg, maskReg, maskNode, loadOpcode, length);
+         resReg = maskReg;
+         }
+      else if (srcNode->getReferenceCount() == 1 && srcNode->getOpCode().isMemoryReference() && srcNode->getRegister() == NULL)
+         {
+         srcReg = cg->allocateRegister();
+         TR::LoadStoreHandler::generateLoadNodeSequence(cg, srcReg, srcNode, loadOpcode, length);
+         resReg = srcReg;
+         }
+      // If no zero-extension shortcut applied, default to zero-extending the mask
+      if (resReg == NULL)
+         {
+         maskReg = cg->gprClobberEvaluate(maskNode);
+         generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rldicl, node, maskReg, maskReg, 0, zeroExtendMask);
+         resReg = maskReg;
+         }
+      // Evaluate the missing operand
+      if (maskReg == NULL)
+         {
+         maskReg = cg->evaluate(maskNode);
+         }
+      else
+         {
+         srcReg = cg->evaluate(srcNode);
+         }
+      }
+   else
+      {
+      srcReg = cg->evaluate(srcNode);
+      maskReg = cg->gprClobberEvaluate(maskNode);
+      resReg = maskReg;
+      }
+
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::pextd, node, resReg, srcReg, maskReg);
+
+   cg->stopUsingRegister((resReg == maskReg) ? srcReg : maskReg);
+
+   node->setRegister(resReg);
+
+   cg->decReferenceCount(srcNode);
+   cg->decReferenceCount(maskNode);
+
+   return resReg;
+   }
+
+TR::Register*
+OMR::Power::TreeEvaluator::bcompressbitsEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   return compressbitsEvaluator(node, cg, TR::Int8);
+   }
+
+TR::Register*
+OMR::Power::TreeEvaluator::scompressbitsEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   return compressbitsEvaluator(node, cg, TR::Int16);
+   }
+
+TR::Register*
+OMR::Power::TreeEvaluator::icompressbitsEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   return compressbitsEvaluator(node, cg, TR::Int32);
+   }
+
+TR::Register*
+OMR::Power::TreeEvaluator::lcompressbitsEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   return compressbitsEvaluator(node, cg, TR::Int64);
+   }
+
+static TR::Register*
+expandbitsEvaluator(TR::Node *node, TR::CodeGenerator *cg, TR::DataType type)
+   {
+   TR::Node *srcNode = node->getFirstChild();
+   TR::Node *maskNode = node->getSecondChild();
+   TR::Register *srcReg = cg->gprClobberEvaluate(srcNode);
+   TR::Register *maskReg = cg->evaluate(maskNode);
+
```

Comment:
you seemed not concerned with zero-extension as in compressingBits methods. unless bits deposited beyond type-size are taken as irrelevant, you should deal with it as well.

Reply:
I'm assuming bits deposited beyond the type-size are taken as irrelevant

Reply:
@zl-wang, I just wanted to double-check this point with you.  Does the p code generation require the results of `TR::Int8` and `TR::Int16` operations to be sign extended to the size of the register, or will later operations perform the sign extension as required?  Will the bits that are deposited beyond the size of the result type yield unexpected results from later operations?

Reply:
these are bit operations so that it was unclear to me whether the result should be expected to be signed, unsigned, or irrelevant. For typical data handling though, the convention (on p) was: guarantee the sign or zero extension up to the default integer type in register (i.e. 32bit). At present, `iload` on p is still using `lwz`, instead of `lwa`.  for byte/short, on the other hand, they need to be zero or sign extended appropriately (at least for the 32bit guarantee).

Reply:
OK, so I think that means @Spencer-Comin will need to sign-extend the results for `TR::Int8` and `TR::Int16` for Power.

Reply:
> guarantee the sign or zero extension up to the default integer type in register

@zl-wang does this mean (1) extend `TR::Int8` and `TR::Int16` results to 32 bits for both 32 and 64 bit modes, or (2) extend `TR::Int8` and `TR::Int16` results to 32 bits for 32 bit mode and extend `TR::Int8`, `TR::Int16`, and `TR::Int32` bit results to 64 bits for 64 bit mode?

Reply:
simply extend to the register width will fulfill the commitment ... extsb and extsh respectively

Reply:
isn't that a requirement across platforms more or less as well? i don't see this as POWER specific

Reply:
I don't believe other platforms make any guarantee on how the bits of the result register beyond the result size are affected by an evaluator (i.e., it's ok to leave garbage in those bits), @hzongaro @r30shah @knn-k is that correct?

Reply:
> simply extend to the register width will fulfill the commitment ... extsb and extsh respectively

fixed in latest push https://github.com/eclipse-omr/omr/commit/0b8e73ee9129b3f0c4ab6a75ed4e328b8ea96f63

Reply:
what happens in reality on other platforms if you use these results as e.g. array index?

Reply:
I spent some time last week looking at how x86 code generation handles one byte and two values.  It was my impression that the contents of the upper portions of registers won't have an impact, but I didn't do a complete survey, so I very well could be mistaken.

@0xdaryl, do you know the answer?  Should @Spencer-Comin sign extend the results to 32 bits on x86?

Reply:
AArch64 usually handles 8- and 16-bit data (signed/unsigned) by extending them to 32-bit wide when they are in registers.

Reply:
From off-line discussion of this question, @0xdaryl indicated that the x86 code generation does not expect the results of `TR::Int8` and `TR::Int16` operations to have been sign-extended to the full width of a 32-bit register.  If any subsequent operation on those results &mdash; for example, `b2i` or `s2i` &mdash; needs to have a full 32-bit value to operate on a register, that subsequent operation is responsible for sign extending the 8 or 16-bit value.